dev/io.openliberty.org.jboss.resteasy.common/src/org/jboss/resteasy/plugins/server/servlet/ServletUtil.java

```diff
@@ -0,0 +1,181 @@
+import org.jboss.resteasy.specimpl.ResteasyUriInfo.InitData;
+import org.jboss.resteasy.util.HttpHeaderNames;
+import org.jboss.resteasy.util.MediaTypeHelper;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.core.Cookie;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
+ * @version $Revision: 1 $
+ */
+public class ServletUtil
+{
+   private static Map<String, InitData> initDataCache = new ConcurrentHashMap<String, InitData>();      // Liberty change
+    
+   public static ResteasyUriInfo extractUriInfo(HttpServletRequest request, String servletPrefix)
+   {
+      String contextPath = request.getContextPath();
+      if (servletPrefix != null && servletPrefix.length() > 0 && !servletPrefix.equals("/"))
+      {
+         if (!contextPath.endsWith("/") && !servletPrefix.startsWith("/"))
+            contextPath += "/";
+         contextPath += servletPrefix;
+      }
+      String queryString = request.getQueryString();
+      String absolute;
+      StringBuffer requestURL = request.getRequestURL();                                                // Liberty change
+      if (queryString != null && queryString.length() > 0)
+      {
+         absolute = requestURL.append('?').append(queryString).toString();                              // Liberty change
+      }
+      else
+      {
+         absolute = requestURL.toString();                                                              // Liberty change
+      }
+      if (!absolute.contains(contextPath))
+      {
+         String encodedContextPath = Arrays.stream(contextPath.substring(1).split("/"))
+                                           .map(s -> {
+                                              try {
+                                                 return URLEncoder.encode(s, "UTF-8");
+                                                } catch (UnsupportedEncodingException ex) {
+                                                 return s;
+                                              }
+                                             })
+                                           .collect(Collectors.joining("/", "/", ""));
+         if (absolute.contains(encodedContextPath))
+         {
+            absolute = absolute.replace(encodedContextPath, contextPath);
+         }
+      }
+      // Liberty change begin      
+      ResteasyUriInfo myResteasyUriInfo;
+      if(InitData.canBeCached(absolute)) {
+          String cacheKey = requestURL.append(contextPath).toString();
```

Comment:
Now you are going to definitely leak because it will have the query string in the cache key.  The ? + queryString were added to it above.

Reply:
There is a reason that the getCacheKey puts in a delimiter between request url and contextpath.  We should do the same or do a small object that has the two Strings in it with an equals to differential between requestURL + contextRoot being different the requestURL with a content of requestURL+contextRoot in the first scenario with a contextRoot of "".

Reply:
After talking with Vijay and making a bunch of runs, I've settled on the "cacheKey class" approach.
Also brought the ee10 version of ServerUtil into the PR, with identical changes.